lib/model/folder_sendrecv.go

```diff
@@ -1336,23 +1336,35 @@ func (f *sendReceiveFolder) copyByWeakHash(state copyBlocksState, w io.WriterAt)
 		h, offset := finder.Match()
 		sha := sha256.Sum256(buf)
 
-		for _, i := range byWeakhash[h] {
-			if !bytes.Equal(state.blocks[i].Hash, sha[:]) {
+		candidates := byWeakhash[h]
+		for i, blockIdx := range candidates {
+			block := &state.blocks[blockIdx]
+			if !bytes.Equal(block.Hash, sha[:]) {
 				continue
 			}
 
-			_, err = f.limitedWriteAt(w, buf, state.blocks[i].Offset)
+			_, err = f.limitedWriteAt(w, buf, block.Offset)
 			if err != nil {
 				state.fail(errors.Wrap(err, "dst write"))
 				return state, err
 			}
 
-			if offset == state.blocks[i].Offset {
+			if offset == block.Offset {
 				state.copiedFromOrigin()
 			} else {
 				state.copiedFromOriginShifted()
 			}
-			done[i] = true
+			done[blockIdx] = true
+
+			// Remove this block from the candidates for h, and remove
+			// h from the Finder if no candidates for it remain.
+			if len(candidates) == 1 {
+				delete(byWeakhash, h)
+				finder.Remove(h)
+			} else {
+				candidates[i] = candidates[len(candidates)-1]
+				byWeakhash[h] = candidates[:len(candidates)-1]
```

Comment:
If we hit this condition multiple times you'll end up with the same block in multiple places.

```
candidates = [1, 2, 3, 4, 5]

// Hit on 1
candidates[0] = 5 // candidates is now  [5, 2, 3, 4, 5]
byWeakHash[h] = [5, 2, 3, 4]

// No hit on 2

// Hit on 3
candidates[2] = 5 // candidates is now  [5, 2, 5, 4, 5]
byWeakHash[h] = [5, 2, 5, 4] // you get saved by the fact len(candidates) is always len(originalCandidates)-1 at worst, hence you never lose 4.
```

Reply:
There's a break statement right after this, so we jump into the next iteration of the outer loop. That resets candidates for the next h found.

Maybe this can be made more obvious by doing

```
candidates := byWeakhash[h]
found := -1
for i, blockIdx := range candidates {
	if theSha256Matches {
		found = i
		break
	}
}

if found == -1 {
	continue
}

// do the actual write and remove candidate here

Reply:
Ah yes, missed that, this now makes sense but brings me to the next question of why we don't want to try the same h for multiple candidates tho?

Reply:
I've set the PR back to draft status until I figure out how to do this right. I'm not even sure if it's correct anymore.